code/authenticators/SAMLSecurityExtension.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+class SAMLSecurityExtension extends Extension {
+
+	/**
+	 * Will redirect the user directly to the ADFS if:
+	 *
+	 * 1) the 'SAMLAuthenticator' is the default authenticator
+	 * 2) there isn't a GET param showloginform set to 1
+	 * 3) the member is currently logged in
+	 * 4) there are form messages (errors or notices)
+	 *
+	 * @return void
+	 */
+	public function onBeforeSecurityLogin() {
+
+		if(Authenticator::get_default_authenticator() != 'SAMLAuthenticator') {
+			return;
+		}
+
+		// by going to the URL Security/login?showloginform=1 we bypass the auto sign on
+		if(isset($_GET['showloginform']) && $_GET['showloginform'] == 1) {
+			return;
+		}
+
+		// if member is already logged in, don't auto-sign-on, this is most likely because
+		// of unsufficient permissions.
+		$member = Member::currentUser();
+		if($member && $member->exists()) {
+			return;
+		}
+
+		// if there are form messages, don't auto-sign-on, this is most likely because of
+		// login errors / failures or other notices.
+		if(Session::get('FormInfo')) {
+			// since FormInfo can be a "nulled" array, we have to check
+			foreach(Session::get('FormInfo') as $form => $info) {
+				foreach($info as $name => $value) {
+					if($value !== null) {
+						return;
+					}
+				}
+			}
+		}
+
+		$backURL = Session::get('BackURL');
+		if(isset($_REQUEST['BackURL'])) {
```

Comment:
FYI, a lot of existing code, e.g. in `MemberLoginForm` doesn't actually adhere to this either. :(